applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py

```diff
@@ -142,34 +157,83 @@ def __analyzeShape(self):
+            if estimation_error < self.approximation_tolerance:
+                new_a = min(current_a*self.increase_factor, self.max_step_size)
+
+            # Leave step size unchanged if a nonliner change in the objective is observed but still a descent direction is obtained
+            else:
+                new_a = current_a
+        else:
+            # Search approximation of optimal step using interpolation
+            a = current_a
+            corrected_step_size = - 0.5 * dfda1 * a**2 / (f2 - f1 - dfda1 * a )
+
+            # Starting from the new design, and assuming an opposite gradient direction, the step size to the approximated optimum behaves reciprocal
+            new_a = current_a-corrected_step_size
+
+        self.step_size = new_a
+
     # --------------------------------------------------------------------------
     def __computeShapeUpdate(self):
         self.mapper.Update()
         self.mapper.InverseMap(DF1DX, DF1DX_MAPPED)
 
         self.optimization_utilities.ComputeSearchDirectionSteepestDescent()
-        self.optimization_utilities.ComputeControlPointUpdate()
+        self.optimization_utilities.ComputeControlPointUpdate(self.step_size)
 
         self.mapper.Map(CONTROL_POINT_UPDATE, SHAPE_UPDATE)
         self.model_part_controller.DampNodalVariableIfSpecified(SHAPE_UPDATE)
 
     # --------------------------------------------------------------------------
     def __logCurrentOptimizationStep(self):
+        self.previos_objective_value = self.communicator.getStandardizedValue(self.objectives[0]["identifier"].GetString())
+        self.norm_obj_gradient = self.optimization_utilities.ComputeL2NormOfNodalVariable(DF1DX_MAPPED)
+
         additional_values_to_log = {}
-        additional_values_to_log["step_size"] = self.algorithm_settings["line_search"]["step_size"].GetDouble()
+        additional_values_to_log["step_size"] = self.step_size
+        additional_values_to_log["norm_obj_gradient"] = self.norm_obj_gradient
         self.data_logger.LogCurrentValues(self.optimization_iteration, additional_values_to_log)
         self.data_logger.LogCurrentDesign(self.optimization_iteration)
 
     # --------------------------------------------------------------------------
     def __isAlgorithmConverged(self):
 
         if self.optimization_iteration > 1 :
-
             # Check if maximum iterations were reached
             if self.optimization_iteration == self.max_iterations:
                 print("\n> Maximal iterations of optimization problem reached!")
                 return True
 
+            # Check gradient norm
+            if self.optimization_iteration == 2:
+                self.initial_norm_obj_gradient = self.norm_obj_gradient
+            else:
+                if self.norm_obj_gradient < self.gradient_tolerance*self.initial_norm_obj_gradient:
+                    print("\n> Optimization problem converged as gradient norm reached specified tolerance of ",self.gradient_tolerance)
```

Comment:
I do not see the advantage of using the Logger for such a simple print statement. Why would you want to change that?

Reply:
ideally all the prints in Kratos should use the logger
The big advantage is that then you can easily filter, redirect, format and suppress messages in a unified way

if you don't use the logger it will just always print, independent if you want the prints or not
e.g. in a test you don't care abt them, you just want to know if the test passes or not